File: engine_py/tests/test_ui_api.py
import pytest

import engine_py


def test_create_button_widget():
    ui = engine_py.UiApi()
    widget_id = ui.create_widget(
        "Button", {"label": "OK", "pos": [1, 2], "color": [255, 255, 255]}
    )
    assert widget_id is not None
    assert widget_id > 0


def test_load_json_button():
    ui = engine_py.UiApi()
    ids = ui.load_json(
        '{"type": "Button", "props": {"label": "Root", "pos": [0,0], "color":'
        " [255,255,255]}}"
    )
    assert isinstance(ids, list)
    assert len(ids) >= 1
    assert ids[0] > 0


def test_remove_widget():
    ui = engine_py.UiApi()
    widget_id = ui.create_widget(
        "Button", {"label": "RemoveMe", "pos": [0, 0], "color": [10, 10, 10]}
    )
    assert widget_id > 0
    removed = ui.remove_widget(widget_id)
    assert removed is True
    removed_again = ui.remove_widget(widget_id)
    assert removed_again is False
    assert ui.remove_widget(999999999) is False


def test_set_widget_props():
    ui = engine_py.UiApi()
    widget_id = ui.create_widget(
        "Button", {"label": "Old", "pos": [0, 0], "color": [1, 2, 3]}
    )
    assert widget_id > 0
    updated = ui.set_widget_props(
        widget_id, {"label": "New", "pos": [5, 6], "color": [7, 8, 9]}
    )
    assert updated is True
    assert ui.set_widget_props(999999999, {"label": "X"}) is False


def test_get_widget_props():
    ui = engine_py.UiApi()
    widget_id = ui.create_widget(
        "Button", {"label": "QueryMe", "pos": [3, 4], "color": [10, 20, 30]}
    )
    props = ui.get_widget_props(widget_id)
    print("props:", props)
    print("type(props):", type(props))
    if props is not None:
        print("props keys:", list(props.keys()))
    assert props is not None
    assert props["label"] == "QueryMe"
    assert isinstance(props["pos"], list)
    assert props["pos"][0] == 3
    assert props["pos"][1] == 4


def test_add_remove_child_panel():
    ui = engine_py.UiApi()
    panel_id = ui.create_widget("Panel", {"pos": [0, 0]})
    btn_id = ui.create_widget(
        "Button", {"label": "Child", "pos": [1, 1], "color": [1, 2, 3]}
    )
    assert ui.add_child(panel_id, btn_id)
    # Now remove it
    assert ui.remove_child(panel_id, btn_id)
    # Remove again should fail
    assert not ui.remove_child(panel_id, btn_id)


def test_set_callback_and_trigger(monkeypatch):
    ui = engine_py.UiApi()
    widget_id = ui.create_widget(
        "Button", {"label": "CB", "pos": [1, 2], "color": [1, 2, 3]}
    )
    called = {"flag": False}

    def cb(wid):
        print("PYTHON: callback called with wid", wid)
        called["flag"] = wid == widget_id

    ui.set_callback(widget_id, "click", cb)
    ui.trigger_event(widget_id, "click", {"x": 1, "y": 2})
    assert called["flag"]


def test_focus_widget():
    ui = engine_py.UiApi()
    widget_id = ui.create_widget(
        "Button", {"label": "Focus", "pos": [1, 2], "color": [1, 2, 3]}
    )
    assert ui.focus_widget(widget_id)


def test_send_ui_event_alias():
    ui = engine_py.UiApi()
    widget_id = ui.create_widget(
        "Button", {"label": "SendEvent", "pos": [1, 2], "color": [1, 2, 3]}
    )
    called = {"flag": False}

    def cb(wid):
        print("PYTHON: callback called with wid", wid)
        called["flag"] = wid == widget_id

    ui.set_callback(widget_id, "click", cb)
    ui.send_ui_event(widget_id, "click", {"x": 1, "y": 2})
    assert called["flag"]


def test_get_widget_type():
    ui = engine_py.UiApi()
    btn_id = ui.create_widget(
        "Button", {"label": "Test", "pos": [0, 0], "color": [255, 255, 255]}
    )
    wtype = ui.get_widget_type(btn_id)
    assert wtype == "Button"

    panel_id = ui.create_widget("Panel", {"pos": [0, 0]})
    wtype = ui.get_widget_type(panel_id)
    assert wtype == "Panel"

    # Non-existent widget returns None or empty string
    assert ui.get_widget_type(999999) in (None, "")


def test_get_parent_and_child_relationship():
    ui = engine_py.UiApi()
    panel_id = ui.create_widget("Panel", {"pos": [0, 0]})
    btn_id = ui.create_widget(
        "Button", {"label": "Child", "pos": [1, 1], "color": [1, 2, 3]}
    )

    # Initially no parent
    assert ui.get_parent(btn_id) is None

    # Add child
    assert ui.add_child(panel_id, btn_id)

    # Parent should be panel_id now
    assert ui.get_parent(btn_id) == panel_id

    # Remove child
    assert ui.remove_child(panel_id, btn_id)

    # Parent should be None again
    assert ui.get_parent(btn_id) is None


def test_set_and_get_z_order():
    ui = engine_py.UiApi()
    btn_id = ui.create_widget(
        "Button", {"label": "ZOrder", "pos": [0, 0], "color": [255, 255, 255]}
    )

    # Default z-order
    assert ui.get_z_order(btn_id) == 0

    # Set z-order
    assert ui.set_z_order(btn_id, 5)

    # Get updated z-order
    assert ui.get_z_order(btn_id) == 5

    # Non-existent widget returns default or error
    assert ui.get_z_order(999999) == 0
    assert not ui.set_z_order(999999, 10)


def test_dynamic_widget_registration():
    ui = engine_py.UiApi()

    # Register a new widget type "CustomWidget" dynamically
    def custom_widget_ctor(props):
        # For test, create a Button with label "Custom"
        return ui.create_widget("Button", props)

    assert ui.register_widget("CustomWidget", custom_widget_ctor)

    # Create widget of new type
    custom_id = ui.create_widget(
        "CustomWidget", {"label": "Dynamic", "pos": [0, 0], "color": [1, 1, 1]}
    )
    assert custom_id > 0

    # Check type is "CustomWidget"
    assert ui.get_widget_type(custom_id) == "CustomWidget"


def test_remove_callback():
    ui = engine_py.UiApi()
    btn_id = ui.create_widget(
        "Button", {"label": "CB", "pos": [1, 2], "color": [1, 2, 3]}
    )
    called = {"flag": False}

    def cb(wid):
        called["flag"] = True

    ui.set_callback(btn_id, "click", cb)
    ui.trigger_event(btn_id, "click", {"x": 1, "y": 2})
    assert called["flag"]

    # Remove callback
    ui.remove_callback(btn_id, "click")

    called["flag"] = False
    ui.trigger_event(btn_id, "click", {"x": 1, "y": 2})
    assert not called["flag"]


def test_get_children_explicit():
    ui = engine_py.UiApi()
    panel_id = ui.create_widget("Panel", {"pos": [0, 0]})
    btn1_id = ui.create_widget(
        "Button", {"label": "Child1", "pos": [1, 1], "color": [1, 2, 3]}
    )
    btn2_id = ui.create_widget(
        "Button", {"label": "Child2", "pos": [2, 2], "color": [4, 5, 6]}
    )

    assert ui.add_child(panel_id, btn1_id)
    assert ui.add_child(panel_id, btn2_id)

    children = ui.get_children(panel_id)
    assert btn1_id in children
    assert btn2_id in children

    # Remove one child
    assert ui.remove_child(panel_id, btn1_id)
    children = ui.get_children(panel_id)
    assert btn1_id not in children
    assert btn2_id in children

File: engine/scripts/lua/tests/test_ui_api.lua
local assert = require("assert")

local function test_create_button_widget()
	local id = ui.create_widget("Button", { label = "OK", pos = { 1, 2 }, color = { 255, 255, 255 } })
	assert.not_nil(id, "ui.create_widget should return a widget id")
	assert.is_true(id > 0, "widget id should be positive")
end

local function test_load_json_button()
	local ids = ui.load_json([[{
        "type": "Button",
        "props": { "label": "Root", "pos": [0, 0], "color": [255, 255, 255] }
    }]])
	assert.not_nil(ids, "ui.load_json should return a table of ids")
	assert.is_table(ids, "ui.load_json should return a table")
	assert.is_true(#ids >= 1, "ui.load_json should return at least one id")
end

local function test_remove_widget()
	local id = ui.create_widget("Button", { label = "RemoveMe", pos = { 0, 0 }, color = { 10, 10, 10 } })
	assert.is_true(id > 0, "Widget id should be positive")

	-- Remove the widget, should succeed
	local removed = ui.remove_widget(id)
	assert.is_true(removed, "Widget should be removed")

	-- Removing again should return false
	local removed_again = ui.remove_widget(id)
	assert.is_false(removed_again, "Widget should not be removed twice")

	-- Removing a random id should return false
	assert.is_false(ui.remove_widget(999999999), "Non-existent widget should not be removed")
end

local function test_set_widget_props()
	local id = ui.create_widget("Button", { label = "Old", pos = { 0, 0 }, color = { 1, 2, 3 } })
	assert.is_true(id > 0, "Widget id should be positive")

	-- Update properties
	local updated = ui.set_widget_props(id, { label = "New", pos = { 5, 6 }, color = { 7, 8, 9 } })
	assert.is_true(updated, "Widget should be updated")

	-- Update non-existent widget
	assert.is_false(ui.set_widget_props(999999999, { label = "X" }), "Non-existent widget should not be updated")
end

local function test_get_widget_props()
	local id = ui.create_widget("Button", { label = "QueryMe", pos = { 3, 4 }, color = { 10, 20, 30 } })
	local props = ui.get_widget_props(id)
	assert.not_nil(props, "get_widget_props should return a table")
	assert.equals(props.label, "QueryMe", "Label should match")
	assert.is_table(props.pos, "Pos should be a table")
	assert.equals(props.pos[1], 3)
	assert.equals(props.pos[2], 4)
end

local function test_add_remove_child_panel()
	local panel_id = ui.create_widget("Panel", { pos = { 0, 0 } })
	local btn_id = ui.create_widget("Button", { label = "Child", pos = { 1, 1 }, color = { 1, 2, 3 } })
	assert.is_true(ui.add_child(panel_id, btn_id), "Should add child to panel")
	-- Now remove it
	assert.is_true(ui.remove_child(panel_id, btn_id), "Should remove child from panel")
	-- Remove again should fail
	assert.is_false(ui.remove_child(panel_id, btn_id), "Should not remove child twice")
end

local function test_set_callback_and_trigger()
	local id = ui.create_widget("Button", { label = "CB", pos = { 1, 2 }, color = { 1, 2, 3 } })
	local called = false
	ui.set_callback(id, "click", function(widget_id)
		called = widget_id == id
	end)
	ui.trigger_event(id, "click", { x = 1, y = 2 })
	assert.is_true(called, "Callback should be called on click event")
end

local function test_focus_widget()
	local id = ui.create_widget("Button", { label = "Focus", pos = { 1, 2 }, color = { 1, 2, 3 } })
	assert.is_true(ui.focus_widget(id), "Should be able to focus widget")
end

return {
	test_create_button_widget = test_create_button_widget,
	test_load_json_button = test_load_json_button,
	test_remove_widget = test_remove_widget,
	test_set_widget_props = test_set_widget_props,
	test_get_widget_props = test_get_widget_props,
	test_add_remove_child_panel = test_add_remove_child_panel,
	test_set_callback_and_trigger = test_set_callback_and_trigger,
	test_focus_widget = test_focus_widget,
}
